lib/Sema/CSSimplify.cpp

```diff
@@ -1668,10 +1670,46 @@ ConstraintSystem::matchDeepEqualityTypes(Type type1, Type type2,
       return result;
   }
 
-  // Match up the generic arguments, exactly.
+  bool forRequirement = false;
+  if (auto last = locator.last()) {
+    forRequirement =
+        last->isTypeParameterRequirement() || last->isConditionalRequirement();
+  }
+  // Optionals and arrays have a lot of special diagnostics and only one
+  // generic argument so if we' re dealing with one, don't produce generic
+  // arguments mismatch fixes.
+  // TODO(diagnostics): Move Optional and Array diagnostics over to the
+  // new framework.
+  bool isOptional = bound1->getDecl()->isOptionalDecl();
+  bool isArray = isArrayType(bound1).hasValue();
+
   auto args1 = bound1->getGenericArgs();
   auto args2 = bound2->getGenericArgs();
-  return matchDeepTypeArguments(*this, subflags, args1, args2, locator);
+
+  SmallVector<unsigned, 4> failedArguments;
+  // Match up the generic arguments, exactly.
+  auto result =
+      matchDeepTypeArguments(*this, subflags, args1, args2, locator,
+                             [&](unsigned position) {
+                                if (forRequirement || isOptional || isArray)
+                                  return false;
+                                
+                                failedArguments.push_back(position);
+                                return true;
+                             });
+  if (!failedArguments.empty()) {
+    auto *fix = GenericArgumentsMismatch::create(
+        *this, bound1, bound2, failedArguments, getConstraintLocator(locator));
+    // Increase the solution's score for each mismtach this fixes.
+    for (unsigned i = 0; i < failedArguments.size(); ++i) {
+      increaseScore(SK_Fix);
```

Comment:
This could be simplified to `increaseScore(SK_Fix, failedArguments.size());`

Reply:
Also I suggest to increase score only if the fix has been recorded successfully e.g.
```c++
if (!recordFix(fix)) {
  increaseScore(SK_Fix, failedArguments.size());
  return getTypeMatchSuccess();
}
```